File: test/dynamic/native/src/js/arrays.rs
use neon::prelude::*;

pub fn return_js_array(mut cx: FunctionContext) -> NeonResult<&JsArray> {
    Ok(cx.empty_array())
}

pub fn return_js_array_with_number(mut cx: FunctionContext) -> NeonResult<&JsArray> {
    let array: &JsArray = JsArray::new(&mut cx, 1);
    let n = cx.number(9000.0);
    array.set(&mut cx, 0, n)?;
    Ok(array)
}

pub fn return_js_array_with_string(mut cx: FunctionContext) -> NeonResult<&JsArray> {
    let array: &JsArray = JsArray::new(&mut cx, 1);
    let s = cx.string("hello node");
    array.set(&mut cx, 0, s)?;
    Ok(array)
}

File: test/dynamic/lib/arrays.js
var addon = require("../native");
var assert = require("chai").assert;

describe("JsArray", function () {
  it("return a JsArray built in Rust", function () {
    assert.deepEqual([], addon.return_js_array());
  });

  it("return a JsArray with a number at index 0", function () {
    assert.deepEqual([9000], addon.return_js_array_with_number());
  });

  it("return a JsArray with an string at index 0", function () {
    assert.deepEqual(["hello node"], addon.return_js_array_with_string());
  });
});
